[YARN] SPARK-6470.  Add support for YARN node labels.

sryza opened the pull request:
This is difficult to write a test for because it relies on the latest version of YARN, but I verified manually that the patch does pass along the label expression on this version and containers are successfully launched.

sryza committed:
[YARN] SPARK-6470.  Add support for YARN node labels.

SparkQA commented:
[Test build #29349 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29349/consoleFull) for   PR 5242 at commit [`7e9c424`](https://github.com/apache/spark/commit/7e9c4246946c8d93be714e28fe913db820cb024c).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29349 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29349/consoleFull) for   PR 5242 at commit [`7e9c424`](https://github.com/apache/spark/commit/7e9c4246946c8d93be714e28fe913db820cb024c).
 * This patch **fails to build**.

 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29349/
Test FAILed.

srowen commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -38,6 +38,7 @@ import org.apache.spark.{SparkEnv, Logging, SecurityManager, SparkConf}
 import org.apache.spark.deploy.yarn.YarnSparkHadoopUtil._
 import org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend
 import org.apache.spark.util.AkkaUtils
+import java.lang.reflect.Constructor
```
Nit, import ordering here.

This looks like a pretty straightforward change, yeah. Too bad this requires reflection, but it's clean. I look forward to being able to assume later versions of YARN / Hadoop to undo some of this. (Although requiring 2.6 probably won't happen too soon.)

The key use case this enables is, for example, running a Spark app on a subset of a cluster's machines that are labeled as having a GPU. I like this given the relatively low impact simple change and the upside to what it enables.

SparkQA commented:
[Test build #29408 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29408/consoleFull) for   PR 5242 at commit [`ce82383`](https://github.com/apache/spark/commit/ce82383cd3ca938c30ed5a5f07260dae77480c7f).

SparkQA commented:
[Test build #29408 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29408/consoleFull) for   PR 5242 at commit [`ce82383`](https://github.com/apache/spark/commit/ce82383cd3ca938c30ed5a5f07260dae77480c7f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29408/
Test PASSed.

tgravescs commented:
mostly looks good.  Seems like we may want to add this to the ApplicationSubmissionContext also to control where the AM can go.  And the question there is do we want that to be able to go to different label then executors.  If so we need to configs.

tgravescs commented:
public static ApplicationSubmissionContext newInstance(
      ApplicationId applicationId, String applicationName, String queue,
      Priority priority, ContainerLaunchContext amContainer,
      boolean isUnmanagedAM, boolean cancelTokensWhenComplete,
      int maxAppAttempts, Resource resource, String applicationType,
      boolean keepContainers, String appLabelExpression,
      String amContainerLabelExpression) 


I think we could potentially use the appLabelExpression for the executors also without having to set it on each ResourceRequest.

leftnoteasy commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -117,6 +117,24 @@ private[yarn] class YarnAllocator(
+  // reflection because it's only available in later versions of YARN.
+  private val nodeLabelConstructor = labelExpression.flatMap { expr =>
+    try {
+      Some(classOf[ContainerRequest].getConstructor(classOf[Resource],
```
@sryza, I think the constructor parameter here is not correct, it should be either 
```
  public static ApplicationSubmissionContext newInstance(
      ApplicationId applicationId, String applicationName, String queue,
      Priority priority, ContainerLaunchContext amContainer,
      boolean isUnmanagedAM, boolean cancelTokensWhenComplete,
      int maxAppAttempts, Resource resource, String applicationType,
      boolean keepContainers, String appLabelExpression,
      String amContainerLabelExpression)
```

Or
```
  public static ApplicationSubmissionContext newInstance(
      ApplicationId applicationId, String applicationName, String queue,
      ContainerLaunchContext amContainer, boolean isUnmanagedAM,
      boolean cancelTokensWhenComplete, int maxAppAttempts,
      String applicationType, boolean keepContainers,
      String appLabelExpression, ResourceRequest resourceRequest) 
```
In 2.6.0 release or after.

vanzin replied:
Looks correct to me. From the hadoop code:

    public ContainerRequest(Resource capability, String[] nodes,
        String[] racks, Priority priority, boolean relaxLocality) {
      this(capability, nodes, racks, priority, relaxLocality, null);
    }

tgravescs commented:
@sryza could you update to do application master as well?

srowen commented:
@sryza are you still looking at this one? it's a good feature. This guy needs a rebase and Thomas has a decent suggestion here too.

sryza commented:
@srowen @tgravescs sorry for the delay here.  Have been working on a new patch that adds support or AM label expressions, but ran into some YARN API issues that makes it difficult.  Will try to post something soon.

sryza commented:
Updated patch still only supports executor node labels, but renames the property so that adding driver/am node labels in the future won't look weird.  I also switched to the API that sets the label expression upon app submission.

SparkQA commented:
[Test build #30252 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30252/consoleFull) for   PR 5242 at commit [`a892793`](https://github.com/apache/spark/commit/a892793535a4e68bb982fe55caf4c0e7bc577960).

SparkQA commented:
[Test build #30252 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30252/consoleFull) for   PR 5242 at commit [`a892793`](https://github.com/apache/spark/commit/a892793535a4e68bb982fe55caf4c0e7bc577960).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30252/
Test PASSed.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -135,6 +136,22 @@ private[spark] class Client(
+        Utils.invoke(classOf[ApplicationSubmissionContext], appContext, "setNodeLabelExpression",
+          (classOf[String], expr))
+      } catch {
+        case e: NoSuchMethodException => logInfo(s"spark.yarn.executor.nodeLabelExpression $expr " +
```
I wonder if this should error out more catastrophically if the user is trying to use a feature that doesn't exist.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -123,6 +123,7 @@ private[spark] class Client(
     appContext.setQueue(args.amQueue)
     appContext.setAMContainerSpec(containerContext)
     appContext.setApplicationType("SPARK")
+    setNodeLabelExpressions(appContext)
```
By setting this on the app launch context, will it also affect the AM?

sryza replied:
I thought I had checked this, but it looks like in fact it will.  Reverting to the old patch.

sryza commented:
It turns out that setting the label expression when creating the `ApplicationSubmissionContext` sets it for the AM as well, which is undesirable.  Updated patch reverts to the original approach of setting the label expression on every container request.

SparkQA commented:
[Test build #30284 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30284/consoleFull) for   PR 5242 at commit [`4a1b837`](https://github.com/apache/spark/commit/4a1b8371fa8156a09c85b372a4c0f279240ffcdf).

SparkQA commented:
[Test build #30284 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30284/consoleFull) for   PR 5242 at commit [`4a1b837`](https://github.com/apache/spark/commit/4a1b8371fa8156a09c85b372a4c0f279240ffcdf).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.
 * This patch **adds the following new dependencies:**
   * `commons-math3-3.1.1.jar`
   * `snappy-java-1.1.1.6.jar`

 * This patch **removes the following dependencies:**
   * `commons-math3-3.4.1.jar`
   * `snappy-java-1.1.1.7.jar`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30284/
Test PASSed.

SparkQA commented:
[Test build #30323 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30323/consoleFull) for   PR 5242 at commit [`e377ed6`](https://github.com/apache/spark/commit/e377ed61e398dbbbda976ba2e61eb0c8488f4c7f).

SparkQA commented:
[Test build #30323 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30323/consoleFull) for   PR 5242 at commit [`e377ed6`](https://github.com/apache/spark/commit/e377ed61e398dbbbda976ba2e61eb0c8488f4c7f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch **adds the following new dependencies:**
   * `commons-math3-3.4.1.jar`
   * `snappy-java-1.1.1.7.jar`

 * This patch **removes the following dependencies:**
   * `commons-math3-3.1.1.jar`
   * `snappy-java-1.1.1.6.jar`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30323/
Test FAILed.

sryza commented:
retest this please

SparkQA commented:
[Test build #30329 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30329/consoleFull) for   PR 5242 at commit [`e377ed6`](https://github.com/apache/spark/commit/e377ed61e398dbbbda976ba2e61eb0c8488f4c7f).

SparkQA commented:
[Test build #30329 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30329/consoleFull) for   PR 5242 at commit [`e377ed6`](https://github.com/apache/spark/commit/e377ed61e398dbbbda976ba2e61eb0c8488f4c7f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30329/
Test PASSed.

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala:
```diff
@@ -231,6 +249,17 @@ private[yarn] class YarnAllocator(
+   * added in recent versions.
+   */
+  private def createContainerRequest(resource: Resource): ContainerRequest = {
+    nodeLabelConstructor.map { constructor =>
```
Just reviving my previous question: should we blow up if the user requests a node label but the library doesn't support the feature?

sryza replied:
I don't have a strong opinion on this, but if you do, I'm happy to change it.

tgravescs replied:
I don't have a strong opinion on this either.  I don't think we've really set a precedence yet. Right now the code matches the documentation.  We could put it in as is (or change logInfo to logWarn) and if we find its an issue change it to error out.   @vanzin  thoughts?

vanzin replied:
I'm fine with making it a `logWarning` just so that it sticks out more in the logs.

vanzin commented:
Looks OK, ignoring the AM-related discussion, but I left a minor question above.

tgravescs commented:
so in the applicationsubmissioncontext it has to label settings:
 String appLabelExpression,
String amContainerLabelExpression

I thought the second was for am and the first for executors, is that not the case then?

vanzin commented:
Hmm... looks like that from a second look at the code. From the code it seems like you could easily do this also:

    appContext.getAMContainerResourceRequest().setNodeLabelExpression(amNodeLabelExpression)

sryza commented:
@tgravescs the first will apply to the AM as well if there's not one for the AM specifically set:
https://github.com/apache/hadoop/blob/trunk/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/RMAppManager.java#L380

@vanzin yes, came across that as well.  My concern was mainly about corner cases with how these get instantiated - do we need to check getAMContainerResourceRequest for null or does it get lazily instantiated.  Not big deals at all, but my bandwidth is low right now, and I'm not convinced that we should add AM node label expressions until we hear of a use case for them, so wanted to put that off for a separate JIRA.

tgravescs commented:
@sryza so to clarify this pr is only going to do this for executor?  is there a jira for doing it for the application master?   Based on above is this ready for review?

sryza commented:
That's correct.  Only for executor, but I updated the config naming after your earlier comment so that we have room to add it for the AM in the future.  I just filed SPARK-7173 for this.

This is ready for review.

SparkQA commented:
[Test build #31037 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31037/consoleFull) for   PR 5242 at commit [`e377ed6`](https://github.com/apache/spark/commit/e377ed61e398dbbbda976ba2e61eb0c8488f4c7f).

tgravescs commented:
@sryza   I think the only remaining comment is to change logInfo to logWarn, otherwise looks good.

tgravescs commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32395 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32395/consoleFull) for   PR 5242 at commit [`e377ed6`](https://github.com/apache/spark/commit/e377ed61e398dbbbda976ba2e61eb0c8488f4c7f).

SparkQA commented:
[Test build #32395 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32395/consoleFull) for   PR 5242 at commit [`e377ed6`](https://github.com/apache/spark/commit/e377ed61e398dbbbda976ba2e61eb0c8488f4c7f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32395/
Test PASSed.

sryza committed:
Change info to warning

sryza commented:
Sorry, missed that.  Just updated the patch.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

srowen commented:
LGTM pending tests. You can tap it in when you're ready.

SparkQA commented:
[Test build #32408 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32408/consoleFull) for   PR 5242 at commit [`6af87b9`](https://github.com/apache/spark/commit/6af87b9b8ab0cc0b79658d29a440e4c975d9b558).

SparkQA commented:
[Test build #32408 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32408/consoleFull) for   PR 5242 at commit [`6af87b9`](https://github.com/apache/spark/commit/6af87b9b8ab0cc0b79658d29a440e4c975d9b558).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32408/
Test PASSed.

sryza commented:
Checked this in.
